pkg/crd/flatten.go

```diff
@@ -26,47 +27,176 @@ import (
 	"sigs.k8s.io/controller-tools/pkg/loader"
 )
 
+// ErrorRecorder knows how to record errors.  It wraps the part of
+// pkg/loader.Package that we need to record errors in places were it might not
+// make sense to have a loader.Package
+type ErrorRecorder interface {
+	// AddError records that the given error occurred.
+	// See the documentation on loader.Package.AddError for more information.
+	AddError(error)
+}
+
+// isOrNil checks if val is nil if val is of a nillable type, otherwise,
+// it comparse val to valuInt (which should probably be the zero value).
+func isOrNil(val reflect.Value, valInt interface{}, zeroInt interface{}) bool {
+	switch valKind := val.Kind(); valKind {
+	case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Ptr, reflect.Slice:
+		return val.IsNil()
+	default:
+		return valInt == zeroInt
+	}
+}
+
 // flattenAllOfInto copies properties from src to dst, then copies the properties
 // of each item in src's allOf to dst's properties as well.
-func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps) {
+func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps, errRec ErrorRecorder) {
 	if len(src.AllOf) > 0 {
 		for _, embedded := range src.AllOf {
-			flattenAllOfInto(dst, embedded)
+			flattenAllOfInto(dst, embedded, errRec)
 		}
 	}
 
-	for propName, prop := range src.Properties {
-		dst.Properties[propName] = prop
+	dstVal := reflect.Indirect(reflect.ValueOf(dst))
+	srcVal := reflect.ValueOf(src)
+	typ := dstVal.Type()
+
+	srcRemainder := apiext.JSONSchemaProps{}
+	srcRemVal := reflect.Indirect(reflect.ValueOf(&srcRemainder))
+	dstRemainder := apiext.JSONSchemaProps{}
+	dstRemVal := reflect.Indirect(reflect.ValueOf(&dstRemainder))
+	hoisted := false
+
+	for i := 0; i < srcVal.NumField(); i++ {
+		fieldName := typ.Field(i).Name
+		if fieldName == "AllOf" {
+			continue
+		}
+		srcField := srcVal.Field(i)
+		fldTyp := srcField.Type()
+		zeroVal := reflect.Zero(fldTyp)
+		zeroInt := zeroVal.Interface()
+		srcInt := srcField.Interface()
+
+		if isOrNil(srcField, srcInt, zeroInt) {
+			// nothing to copy from src, continue
+			continue
+		}
+
+		dstField := dstVal.Field(i)
+		dstInt := dstField.Interface()
+		if isOrNil(dstField, dstInt, zeroInt) {
+			// dst is empty, continue
+			dstField.Set(srcField)
```

Comment:
this does not work for nullable.

```
type: array
allOf:
- type: array
   nullable: true
```
is not equivalent to
```
type: array
nullable: true
```
because `null` is valid for the latter, but not for the former.

Reply:
I think with structural schemas we are saved here: nullable is not allowed below allOf.

Reply:
huh?  How does that work?